### ISSUE TITLE
why requantization output  need add  src_zero_point*src_zero_point_compenzation?

### ISSUE THREAD

**Issue by Daniel-NJ** [oneapi-src/oneDNN#1024]
i currently read this kernel, src\cpu\x64\jit_uni_x8s8s32x_conv_kernel.cpp
in function below it requantize the s32 to u8/s8, according the quantization model, s8 = s32*scale_out + output_zero_point, why
need add zp_comp? could you pls give some formula for these process， thanks！

>  template <cpu_isa_t isa, typename Vmm>
void _jit_uni_x8s8s32x_fwd_kernel<isa, Vmm>::store_output(
.......
        for (int j = 0; j < ur_w; ++j) {
            const Vmm vmm = vmm_out(j, k);

            /* add comp in s32 to avoid loss of precision
               when convert s32 to f32 in integer (2^24)
               TODO: do the same to bias */
            if (jcp.signed_input) uni_vpaddd(vmm, vmm, vmm_comp);
            if (jcp.src_zero_point) uni_vpaddd(vmm, vmm, vmm_zp_comp);//cannot understand what these means?
            uni_vcvtdq2ps(vmm, vmm);

**Comment by mgouicem** [comment]
Hi @Daniel-NJ , this is a very good question!
We assume that each quantized tensor is represented as follow: `val_f32 = (val_x8 - val_zp) * val_scale`
To simplify notation, I will just write down what happens for a dot product, and the same formula will naturally apply to both convolution and matmul. So the dot product formula is:
`dst_f32 = sum_i [ src_f32[i] * wei_f32[i] ]`

After quantization (let's assume no zp for wei), this equation becomes:
`(dst_x8 - dst_zp) * dst_scale = sum_i [ (src_x8[i] - src_zp) * src_scale * wei_s8[i] ]`
`dst_x8 = dst_zp + 1/dst_scale * sum_i [ (src_x8[i] - src_zp) * src_scale * wei_s8[i] ]`

Evaluating the above formula directly would be inneficient (because of the src_zp and src_scale terms). Hence we try to simplify it. Because src_scale is independant of i, we can extract it from the sum:
`dst_x8 = dst_zp + * src_scale/dst_scale * sum_i [ (src_x8[i] - src_zp)  * wei_s8[i] ]`

We can also distribute the product inside the sum:
`dst_x8 = dst_zp + src_scale/dst_scale * sum_i [ (src_x8[i] * wei_s8[i] - src_zp * wei_s8[i] ]`

We split the sum on i with a pure x8 dot product, and another term:
`dst_x8 = dst_zp + src_scale/dst_scale * { sum_i [ (src_x8[i] * wei_s8[i]) ] - sum_i [ src_zp * wei_s8[i] ] }`

We extract src_zp from the sum as it is independent of i
`dst_x8 = dst_zp + src_scale/dst_scale * { sum_i [ (src_x8[i] * wei_s8[i]) ] - src_zp * sum_i [ wei_s8[i] ] }`

It may seem worse than the initial formula, but it is actually much better :-). One thing we can do, is to precompute the compensation (`- sum_i [ wei_s8[i] ]`) as part of reordering the weights. This means that when executing our primitive (convolution or matmul), we actually:
- compute the x8 dot product ` sum_i [ (src_x8[i] * wei_s8[i]) ] ` this can be done very quickly by chaining FMAs or VNNI instructions.
- compute [the product src_zp * compensation](https://github.com/oneapi-src/oneDNN/blob/e397f16d43b91105a4e866b040c72be10bebdc5c/src/cpu/x64/jit_avx512_core_x8s8s32x_conv_kernel.cpp#L282-L284) (with `compensation = - sum_i [ wei_s8[i] ]`). This is fast and happens only once per dot product.
- apply the output_scale = src_scale / dst_scale ([here](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/x64/jit_avx512_core_x8s8s32x_conv_kernel.cpp#L301-L302))
- [add the dst_zp](https://github.com/oneapi-src/oneDNN/blob/e397f16d43b91105a4e866b040c72be10bebdc5c/src/cpu/x64/jit_avx512_core_x8s8s32x_conv_kernel.cpp#L308-L319) 

Hope this helps!

**Comment by Daniel-NJ** [comment]
Hi @mgouicem , thanks much for your so quick feedback.  your answer is very helpful for me, there are another question, dose the scale and src_zp are per channel parameters?

**Comment by Daniel-NJ** [comment]
Hi @mgouicem，i wander whether in oneDNN, the scale is use  KLD (KL divergence) method to find the threshold T(per channel)， scale = 127/T,  but how you set up zero_point per tensor?  thanks.

**Comment by mgouicem** [comment]
How those scales and zero point is out of scope of oneDNN and are computed by the oneDNN user.
With that said, there are various ways of computing those quantization parameters and we generally recommend to rely on quantization tools, like Intel [LPOT](https://github.com/intel/lpot) or Intel [OpenVINO](https://docs.openvinotoolkit.org/latest/pot_docs_LowPrecisionOptimizationGuide.html).

The general idea though is to clamp activation in an interval `[a, b]`, and then set the zero point as the midpoint of that interval (`(b - a) / 2`).  That clamping can be determined in many ways, here are some examples:
- it can happen naturally because of the nature of the computation. e.g. softmax and sigmoid output values in `[0; 1]`.
- clamping can be applied during training (when doing quantization aware training).
- clamping can be "learned" post training, by looking at the distribution of activation on a reference dataset.

Hope this helps!

**Comment by Daniel-NJ** [comment]
thanks very much ：）
